bolt/spark/swap.py

```diff
@@ -1,242 +1,356 @@
 
+        sizes = self.plan
+
         def _extract(record):
 
             k, v = record[0], record[1]
 
-            chunk, stationary_key = k[0], k[1]
-            key_offsets = prod([asarray(chunk), asarray(chunk_sizes)[vmask]], axis=0)
-            moving_keys, values = zip(*v.data)
-            sorted_keys = tuplesort([i.tolist() for i in moving_keys])
-            values_sorted = asarray(values)[sorted_keys]
-            expanded_shape = concatenate([moving_key_shape, values_sorted.shape[1:]])
-            bounds = asarray(values_sorted[0].shape)[vmask]
+            stationary_key, chunk = k[0], k[1]
+            key_offsets = prod([asarray(chunk), asarray(sizes)[vmask]], axis=0)
+
+            bounds = asarray(v.shape[len(kaxes):])[vmask]
             indices = list(product(*map(lambda x: arange(x), bounds)))
-            values = values_sorted.reshape(expanded_shape)
 
             for b in indices:
                 s = slices.copy()
                 s[mask] = b
-                yield (tuple(asarray(r_[stationary_key, key_offsets + b], dtype='int')), values[tuple(s)])
-
-        return rdd.flatMap(_extract)
+                yield (tuple(asarray(r_[stationary_key, key_offsets + b], dtype='int')), v[tuple(s)])
+            
+        rdd = rdd.flatMap(_extract)
+        split = self._split - len(kaxes) + len(vaxes)
+        shape = tuple(r_[self.kshape[~kmask], self.vshape[vmask],
+                         self.kshape[kmask], self.vshape[~vmask]].astype('int'))
+        return BoltArraySpark(rdd, shape=shape, split=split)
 
-    def getplan(self, size, dtype):
+    def getplan(self, size="150", axes=None):
         """
-        Identify the plan for chunking along each value-dimension. This
-        generates an ndarray with the number of chunks in each
-        dimension. Any dimension that is staying in the values is set
-        as a single chunk.
+        Identify a plan for chunking values along each dimension.
 
-        Typical size parameter is 150 (an int, megabytes)
+        Generates an ndarray with the size (in number of elements) of chunks
+        in each dimension as well as the original dimensions of the values
+        used in this computation. If provided, will estimate chunks for only a
+        subset of axes, leaving all others to the full size of the axis.
 
         Parameters
         ----------
-        size : integer or tuple
-             If int, the average size of the chunks in all value dimensions.  
-             If tuple, an explicit specification of the number chunks in 
+        size : string or tuple
+             If str, the average size (in MB) of the chunks in all value dimensions.  
+             If int/tuple, an explicit specification of the number chunks in 
              each moving value dimension.
 
-        dtype : dtype 
-              Valid dtype of the underlying data, used to calculate 
-              size in each chunk.
+        axis : tuple, optional, default=None
```

Comment:
`axis` or `axes`? looks like you used `axes` as the actual var and it's just an typo in the docstring.